text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,196 @@
+during compile-time, the engine will assert it during runtime.
+
+### Testing
+
+LWC currently has integration tests for synthetic shadow mode and native shadow mode. A third
+testing mode called "mixed shadow mode" will be introduced. Synthetic shadow DOM polyfills are
+implemented as patches on global prototypes so the decision of invoking a polyfilled API vs a native
+API will depend on the preference of individual components during runtime. This test environment
+will ensure that components designed to run in a native Shadow DOM environment will work correctly
+when even when synthetic Shadow DOM polyfills are present.
+
+Components that prefer native mode will need to also support synthetic mode in browser environments
+where Shadow DOM is natively unavailable, so they will also need to be tested in both modes. LWC
+test utilities will need to be updated to facilitate this.
+
+The existing WPT (Web Platform Tests) test suite for Shadow DOM APIs will also be run to identify
+any coverage gaps in LWC integration tests.
+
+## Observable differences
+
+### Light DOM and assigned elements
+
+In native Shadow DOM, elements are rendered in a component's light DOM and assigned to slots in a
+child component's shadow DOM. This means that elements will exist in the DOM regardless of whether
+they are assigned to a slot. This is not the case for LWC--elements that are passed down to child
+components are never rendered unless they are assigned to a slot.
+
+In the following example, `span` will not exist in the DOM in synthetic shadow, but it will exist in
+native shadow.
+
+```html
+<!-- x-parent -->
+<template>
+    <x-child>
+        <span>foo</span>
+    </x-child>
+</template>
+
+<!-- x-child -->
+<template>
+    <p>child</p>
+</template>
+```
+
+This observable difference will not be remediated as such a change would be non-trivial and
+backwards-incompatible with the current synthetic shadow implementation.
+
+### Lifecycle timing
+
+Due to the way that slotting is implemented in `@lwc/synthetic-shadow`, there is an observable
+difference in the timing of lifecycle hooks for slotted elements.
+
+As mentioned previously, in synthetic mode, slotted elements that are never assigned to a slot are
+not rendered. This means that their lifecycle hooks are never invoked.
+
+Another difference is that for synthetic mode, lifecycle hooks are invoked in the order of
+appearance after they are assigned, whereas in native mode, they are invoked in the order of
+appearance in the template.
+
+### this.shadowRoot vs this.template
+
+`this.shadowRoot` and `this.template` reference the component's shadow root in native Shadow DOM
+while `this.template` references the component's shadow root in synthetic Shadow DOM. There are no
+plans to enable `this.shadowRoot` for existing components that only support synthetic Shadow DOM as
+the current LWC-ism is to use `this.template`.

```

Comment:
Does this mean that, moving forward, we will recommend `this.shadowRoot` over `this.template`? (I think this may actually be a good idea, since one is standard and the other is an LWC-ism.)

Reply:
Actually, this section states that we will not enable `this.shadowRoot` for components unless they prefer native mode.

Reply:
Gotcha. In that case, backporting `this.shadowRoot` to synthetic (in which case it just refers to `this.template`) would seem sensible to me, since it's one less thing for developers to have to think about when converting components between native/synthetic shadow.

Reply:
I would prefer to wait and see if this is an actual pain point before backporting this feature. I suspect it won't be because conversions would just be from synthetic to native and `this.template` would work in both modes. If we do decide to make the change later, it would be a simple and straightforward backwards-compatible change, so we probably won't need to reach consensus to move forward. I will add this as an open question to get some more feedback.

Reply:
Thinking about this more, I think my original intention was to expose `this.shadowRoot` in synthetic mode, iff `preferNativeShadow` is set to `true`. This makes sense to me so I'll make it more explicit in this section:

> There are no plans to enable `this.shadowRoot` for existing components that **_only_** support synthetic Shadow DOM as the current LWC-ism is to use `this.template`.